operator/validator/controller.go

```diff
@@ -986,107 +915,52 @@ func (c *controller) startValidator(v *validator.Validator) (bool, error) {
-			end := remainder
-			if end > len(existingShares) {
-				end = len(existingShares)
-			}
-			shares = append(shares, existingShares[:end]...)
-		}
-		for _, share := range shares {
-			share.SetMetadataLastUpdated(time.Now())
-		}
-
-		filteringTook := time.Since(start)
-		if len(shares) > 0 {
-			pubKeys := make([][]byte, len(shares))
-			for i, s := range shares {
-				pubKeys[i] = s.ValidatorPubKey[:]
-			}
-			err := c.fetchAndUpdateValidatorsMetadata(c.logger, pubKeys, c.beacon)
-			if err != nil {
-				c.logger.Warn("failed to update validators metadata", zap.Error(err))
-				continue
-			}
-		}
-		c.logger.Debug("updated validators metadata",
-			zap.Int("validators", len(shares)),
-			zap.Int("new_validators", len(newShares)),
-			zap.Uint64("started_validators", c.recentlyStartedValidators),
-			zap.Duration("filtering_took", filteringTook),
-			fields.Took(time.Since(start)))
-
-		// Only sleep if there aren't more validators to fetch metadata for.
-		if len(shares) < batchSize {
-			time.Sleep(sleep)
+func (c *controller) HandleMetadataUpdates(ctx context.Context) {
+	for validatorMap := range c.metadataUpdater.Stream(ctx) {
+		if err := c.handleMetadataUpdate(ctx, validatorMap); err != nil {
+			c.logger.Warn("could not handle metadata update", zap.Error(err))
 		}
 	}
 }
 
-func (c *controller) fetchAndUpdateValidatorsMetadata(logger *zap.Logger, pks [][]byte, beacon beaconprotocol.BeaconNode) error {
-	// Fetch metadata for all validators.
-	c.recentlyStartedValidators = 0
-	beforeUpdate := c.AllActiveIndices(c.beacon.GetBeaconNetwork().EstimatedCurrentEpoch(), false)
-
-	err := beaconprotocol.UpdateValidatorsMetadata(logger, pks, beacon, c.UpdateValidatorsMetadata)
-	if err != nil {
-		return errors.Wrap(err, "failed to update validators metadata")
+func (c *controller) handleMetadataUpdate(ctx context.Context, validatorMap metadata.ValidatorMap) error {
+	startedValidators := 0
+	if c.operatorDataStore.GetOperatorID() != 0 {
+		startedValidators = c.startValidatorsForMetadata(ctx, validatorMap)
 	}
 
-	// Refresh duties if there are any new active validators.
-	afterUpdate := c.AllActiveIndices(c.beacon.GetBeaconNetwork().EstimatedCurrentEpoch(), false)
-	if c.recentlyStartedValidators > 0 || hasNewValidators(beforeUpdate, afterUpdate) {
+	if startedValidators > 0 {
 		c.logger.Debug("new validators found after metadata update",
-			zap.Int("before", len(beforeUpdate)),
-			zap.Int("after", len(afterUpdate)),
-			zap.Uint64("started_validators", c.recentlyStartedValidators),
+			zap.Int("started_validators", startedValidators),
 		)
-		select {
-		case c.indicesChange <- struct{}{}:
-		case <-time.After(2 * c.beacon.GetBeaconNetwork().SlotDurationSec()):
+		// Refresh duties if there are any new active validators.
+		if !c.reportIndicesChange(ctx, 2*c.beacon.GetBeaconNetwork().SlotDurationSec()) {
 			c.logger.Warn("timed out while notifying DutyScheduler of new validators")
 		}
```

Comment:
is there a reason we're not checking the indices like before?

iirc this is the reliable check for new indices and the started check is likely legacy from before

Reply:
@moshe-blox I saw this comment https://github.com/ssvlabs/ssv/pull/1805#discussion_r1896912761 and thought that as it happens after starting validators then the most reliable way would be to check how many validators actually started. And if I remove `hasNewValidators`, then the code with checking indices becomes unused

Reply:
i see

well, to answer whether either 1 is enough or not, research is needed

it's best to not remove unless we understand the implications

if i had to guess the former check is more reliable

i'd keep both in any case until we're sure